### ISSUE TITLE
Ladder operators in PauliStrings

### ISSUE THREAD

**Issue by jwnys** [netket/netket#1206]
[This issue is part of [UnitaryHack](https://unitaryhack.dev/) and comes with a bounty of 100$]

**Context**
NetKet implements PauliStrings operators that implement the Pauli matrix sums. For example, one can create an TFIM with 2 spins as
```python
nk.operator.PauliStrings(hilbert, operators=("ZZ", "XI", "IX"), weights=(-J, -h, -h))
```
However, notice that if we would implement ladder operators such as (X+iY)/2, we would need to sum over two terms. This becomes problematic when we implement e.g. Jordan-Wigner transformations which contain ladder operators at multiple sites, such that the number of terms in the sum in the operators blows up. Currently, NetKet therefore implements a custom version of fermionic operators which does the Jordan-Wigner transformation more efficiently. This is clearly not ideal, and there are many more examples where ladder operators become important.

**Implementation notes**
To resolve this issue, we need ladder operators. We can allow the PauliStrings to take strings "+" and "-" for a ladder operator (X+iY)/2 and (X-iY)/2 respectively. The main difference with other Pauli operators (X, Y, Z, I) is that these operators will often generate matrix elements that are zero. Nevertheless, it will avoid the current behavior of NetKet, which regards X and Y separately in the ladder operators. We can of course also extend this to ladder operators that operate on other hilbert space than just Qubit and Spin.

The implementation requires parsing the input strings to detect "+" and "-" gates, and to create connected elements accordingly. More than that, NetKet also allows a researcher to do symbolic manipulations with PauliStrings, and any new type of operator would have to implement this as well. Hence, we need to implement the commutation relations and so on.

Currently, NetKet parses the PauliStrings as follows in order to compute matrix elements for a given set of samples
- X: flips the spin, does not give phase factor
- Y: flips the spin, gives a phase factor
- Z: does not flip the spin, gives a phase factor
- I: does not flip the spin, does not change the phase
We see that we can know in advance which operators affect the "mels" (matrix elements) and which create "connected elements" different from the input element. The "+" and "-" will now need to be parsed as well. However, these now generate 0 matrix elements, and hence, we will need to store the locations of such "ladder operators". Once this is known, we know that it will often happen that the corresponding matrix elements vanish, and that there is therefore not always a connected element for each sample (in contrary to other PauliStrings which always have 1 connected element).

Once we have this implemented, we can do many things like fermionic operators much more efficiently and consistently, which opens up the road to quantum chemistry with NetKet!

**Comment by SakibulIslamSazzad** [comment]
Hi, I am interested in this issue for unitary hack. Can you please guide me on how I can start? Is there any tutorials/documentation to work on this issue?

**Comment by PhilipVinc** [comment]
Hi @SakibulIslamSazzad 

Have you ever used NetKet before? If not I'd advise you to give a read to the section 2.2 of our paper https://arxiv.org/pdf/2112.10526.pdf where the operators are described a bit more than on our website.

In NetKet we have a bunch of different operators which you can think of as sparse matrices which expose a method, `get_conn_flattened/padded` which returns connected elements and matrix elements.

For example, if we consider the `X` operator I can build it as
```python
>>> import netket as nk

>>> op = nk.operator.PauliStrings(["X"])
```
This is an operator acting on a single qubit, so if I check all configurations in the Hilbert space they are exactly `0` and `1`.
```python
>>> all_states = op.hilbert.all_states()
>>> print(all_states)
array([[0.],
       [1.]])
```
As I know that the X operator has a matrix element `1.0` that links `0 -> 1` and `1->0`, I expect that the connected element of `0` will be `1` with matrix element 1, and `1` will be connected to `0` with matrix element `1.0`.
See for example (where `all_states[0]` is `[0]`):

```python
>>> connected_elements, matrix_elements = op.get_conn(all_states[0])

>>> print(connected_elements)
[[1.]]

>>> print(matrix_elements)
[1.+0.j]
```
and (where `all_states[1]` is `[1]`):
```python
>>> connected_elements, matrix_elements = op.get_conn(all_states[1])

>>> print(connected_elements)
[[0.]]

>>> print(matrix_elements)
[1.+0.j]
```

Does it make sense?

Now, if you have a PauliString with `XI`, acting on two sites, but with the identity on the second one, you will see very similar results:
```python
In [17]: op = nk.operator.PauliStrings(["XI"])

In [18]: all_states = op.hilbert.all_states()

In [19]: print(all_states)
[[0. 0.]
 [0. 1.]
 [1. 0.]
 [1. 1.]]

In [20]: connected_elements, matrix_elements = op.get_conn_padded(all_states)

In [21]: for (i, (conn_el, matrix_el)) in enumerate(zip(connected_elements, matrix_elements)):
    ...:     print(f"state: {all_states[i]}")
    ...:     print(f"  connected elements: {conn_el}")
    ...:     print(f"  matrix elements:    {matrix_el}")
    ...:
state: [0. 0.]
  connected elements: [[1. 0.]]
  matrix elements:    [1.+0.j]
state: [0. 1.]
  connected elements: [[1. 1.]]
  matrix elements:    [1.+0.j]
state: [1. 0.]
  connected elements: [[0. 0.]]
  matrix elements:    [1.+0.j]
state: [1. 1.]
  connected elements: [[0. 1.]]
  matrix elements:    [1.+0.j]
```

wheeee we essentially have the same results as before, but the configurations have an extra qubit that is not changed by the operator.

The heart of the implementation is inside of the [`netket/operator/_pauli_strings.py`](https://github.com/netket/netket/blob/master/netket/operator/_pauli_strings.py) file. Those operators are constructed by passing a list of strings like `[XXX, YXZ]`, etc, and then In the [`_setup`](https://github.com/netket/netket/blob/cfe2971d3455ef597679070fe6f097ef2317dadd/netket/operator/_pauli_strings.py#L132) method we convert those strings into a set of data structures that can be efficiently used to compute connected elements.

Then, when you want to `get_conn`ected elements, all calls eventually end up in a numba function [`_flattened_kernel`](https://github.com/netket/netket/blob/cfe2971d3455ef597679070fe6f097ef2317dadd/netket/operator/_pauli_strings.py#L361), that takes a list of those configurations (`all_states` in the case above) and computes the connected configurations and matrix elements.

NetKet nowadays supports `X,Y,Z,I` operators (all Pauli matrices), but we would like to also support `+` and `-`.

--

Implementation wise,  you should first extend the logic of `_setup` to also handle `+` and `-` in the PauliStrings and store a flag specifying that there's a projector in that site/operator.
If you do the setup properly, you should only slightly modify the `_flattened_kernel` function to set the matrix element to `0` if `+` acts on `1` or if `-` acts on `0`.

--

If you've never used netket's operators before this task might be a bit intense at first while you navigate through this logic, and the unitary hack issue about streaming statistics might be easier for novices.
If you still are set on doing this, I can try to help you while you try to make out all the details of the PauliOperator. 
Anyway, even just adding some comments to the code would be great.

**Comment by gcarleo** [comment]
can we close this? was it implemented in #1279?

**Comment by jwnys** [comment]
We can close this, the original idea was to map all the fermionic operators onto PauliStrings, but that no longer seems like a good idea now

**Comment by gcarleo** [comment]
converting to PauliStrings (at least in the case of LocalOperator) offers a nice simplification / speedup of the matrix elements , I suspect that having a conversion of fermions too might already offer some speedup also for fermions ?

**Comment by jwnys** [comment]
2. Sure
1. Optimizing fermions will just require specific hamiltonians with very specific optimizations... I don't think that this general approach is the best way to go. Currently I don't see much of a speedup wrt to the fermions implementation. You can still implement it and compare, of course.. It will just make PauliStrings a bit more complicated.. And it's harder to estimate the max conn size mainly.

**Comment by jwnys** [comment]
> converting to PauliStrings (at least in the case of LocalOperator) offers a nice simplification / speedup of the matrix elements , I suspect that having a conversion of fermions too might already offer some speedup also for fermions ?

2 complications here wrt why I did it for LocalOperator only: non-locality of fermions and max conn size that is harder to estimate

**Comment by PhilipVinc** [comment]
Ah, yes! I agree wholly!

It's true that implementing ladder operators for PauliStrings would complicate them a bit (but I think the max Conn calculation would then match what you do in fermions) but I think it's a common enough case (tight binding Hamiltonian) that it's worth implementing even if it's not the most performant. I'm unsure if in here or in another operator type..

**Comment by gcarleo** [comment]
indeed, that's why it's important to have a specialized Hamiltonian for T+V interactions (and conserved number of particles), in there we can then push all the specific optimization we need. But OK, I agree that converting to PauliString is not a good idea to speed up, overall. Still, it is maybe useful to have a simple converter of fermionic hamiltonians to PauliString, just to have a common baseline for all hamiltonians with 2 local states

**Comment by jwnys** [comment]
> indeed, that's why it's important to have a specialized Hamiltonian for T+V interactions (and conserved number of particles), in there we can then push all the specific optimization we need. But OK, I agree that converting to PauliString is not a good idea to speed up, overall. Still, it is maybe useful to have a simple converter of fermionic hamiltonians to PauliString, just to have a common baseline for all hamiltonians with 2 local states

Agree on the first, this would be priority for the fermions.
A converter could already be made, you don't necessarily need a + and - operator in PauliStrings... You can write one with just X \pm iY (it will just be way less efficient)
